File: Java/프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        Set<String> set = new HashSet<>(Arrays.asList(gems));
        int total = set.size();
        int start = 0;
        int end = 0;
        Map<String, Integer> map = new HashMap<>();

        int minLength = Integer.MAX_VALUE;
        int answerStart = 0;
        int answerEnd = 0;

        while (true) {
            if (map.size() < total) {
                if (end == gems.length) {
                    break;
                }
                map.put(gems[end], map.getOrDefault(gems[end], 0) + 1);
                end++;
            } else {
                if (end - start < minLength) {
                    minLength = end - start;
                    answerStart = start;
                    answerEnd = end;
                }
                map.put(gems[start], map.get(gems[start]) - 1);
                if (map.get(gems[start]) == 0) {
                    map.remove(gems[start]);
                }
                start++;
            }
        }
        answer[0] = answerStart + 1;
        answer[1] = answerEnd;

        return answer;
    }
}

File: Python/프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    answer = [0, len(gems)]
    size = len(set(gems))
    s = 0
    d = {}
    
    for e in range(len(gems)):
        gem = gems[e]
        d[gem] = d.get(gem, 0) + 1
        
        while size == len(d):
            if e - s < answer[1] - answer[0]:
                answer = [s, e]
            
            left_gem = gems[s]
            d[left_gem] -= 1
            if d[left_gem] == 0:
                del d[left_gem]
            
            s+=1
    
    return [answer[0] + 1, answer[1] + 1]